File: currency-hedge.py
# Fetch the current exchange rate 
from forex_python.converter import CurrencyRates

# Create an instance of CurrencyRates
cr = CurrencyRates()

# Define the currencies you want to convert
base_currency = 'USD'
target_currency = 'EUR'

# Fetch the current exchange rate
exchange_rate = cr.get_rate(base_currency, target_currency)

# Print the exchange rate
print(f"Current exchange rate: 1 {base_currency} = {exchange_rate} {target_currency}")






class CurrencyHedging:
    @staticmethod
    def settle(amount, exchange_rate):
        """
        Calculates the settlement amount in the target currency based on the amount to be hedged and the exchange rate.
        Args:
            amount (float): Amount to be hedged.
            exchange_rate (float): Exchange rate between the base currency and the target currency.
        Returns:
            float: Settlement amount in the target currency.
        """
        return amount * exchange_rate


def currency_hedging(amount, current_rate, forward_rate, transaction_cost, margin_requirement, risk_tolerance,
                     stop_loss=None, take_profit=None, trailing_stop=None, partial_hedging_percentage=None,
                     diversification_factor=None, rebalancing_percentage=None, correlation_threshold=None,
                     volatility_stop_loss=None, profit_target=None, max_drawdown=None, cost_average_factor=None,
                     dynamic_position_multiplier=None, option_cost=None, stress_testing=None, scenario_analysis=None):
    """
    Executes the currency hedging strategy.
    Args:
        amount (float): Amount to be hedged in the base currency.
        current_rate (float): Current exchange rate between the base and target currencies.
        forward_rate (float): Forward exchange rate between the base and target currencies.
        transaction_cost (float): Transaction cost in the target currency.
        margin_requirement (float): Margin requirement as a decimal (e.g., 0.1 = 10%).
        risk_tolerance (float): Risk tolerance in the target currency.
        stop_loss (float, optional): Stop loss threshold. Default is None.
        take_profit (float, optional): Take profit threshold. Default is None.
        trailing_stop (float, optional): Trailing stop loss distance. Default is None.
        partial_hedging_percentage (float, optional): Percentage of hedged amount to be partially hedged. Default is None.
        diversification_factor (float, optional): Proportion of remaining hedged amount allocated for diversification. Default is None.
        rebalancing_percentage (float, optional): Percentage of position size to be rebalanced. Default is None.
        correlation_threshold (float, optional): Correlation threshold for position size adjustment. Default is None.
        volatility_stop_loss (float, optional): Stop loss based on volatility. Default is None.
        profit_target (float, optional): Profit target threshold. Default is None.
        max_drawdown (float, optional): Maximum allowed drawdown as a decimal value. Default is None.
        cost_average_factor (float, optional): Percentage of position size to be cost averaged. Default is None.
        dynamic_position_multiplier (float, optional): Multiplier for dynamic position sizing. Default is None.
        option_cost (float, optional): Cost of options used for hedging. Default is None.
        stress_testing (float, optional): Percentage of stress-tested position size. Default is None.
        scenario_analysis (float, optional): Percentage of scenario-based adjusted position size. Default is None.
    Returns:
        tuple: A tuple containing the settlement amount (float) and the hedged amount (float).
    """
    rate_difference = forward_rate - current_rate

    settlement_amount = CurrencyHedging.settle(amount, forward_rate)

    # Deduct transaction cost from settlement amount
    settlement_amount -= transaction_cost

    # Calculate the margin required based on the hedged amount
    margin_required = amount * margin_requirement

    # Adjust the hedged amount based on the margin requirement
    hedged_amount = amount + (amount * rate_difference) - margin_required

    # Apply risk management strategies
    if stop_loss is not None and current_rate <= stop_loss:
        hedged_amount = 0

    if take_profit is not None and current_rate >= take_profit:
        hedged_amount = risk_tolerance

    if trailing_stop is not None and trailing_stop > 0:
        trailing_stop_loss = max(current_rate - trailing_stop, stop_loss)
        if current_rate <= trailing_stop_loss:
            hedged_amount = 0

    if partial_hedging_percentage is not None:
        partial_hedged_amount = hedged_amount * partial_hedging_percentage
        remaining_hedged_amount = hedged_amount - partial_hedged_amount
        hedged_amount = min(hedged_amount, remaining_hedged_amount)

    if diversification_factor is not None:
        diversification_hedged_amount = remaining_hedged_amount * diversification_factor
        remaining_hedged_amount -= diversification_hedged_amount
        hedged_amount = min(hedged_amount, remaining_hedged_amount)

    if rebalancing_percentage is not None:
        rebalancing_amount = hedged_amount * rebalancing_percentage
        hedged_amount -= rebalancing_amount

    if correlation_threshold is not None:
        if correlation_threshold < 0:
            # Negative correlation: Reduce position size
            hedged_amount *= abs(correlation_threshold)
        else:
            # Positive correlation: Increase position size
            hedged_amount *= 1 + correlation_threshold

    if volatility_stop_loss is not None and current_rate - volatility_stop_loss >= stop_loss:
        hedged_amount = 0

    if profit_target is not None and current_rate >= profit_target:
        hedged_amount = risk_tolerance

    if max_drawdown is not None and max_drawdown > 0 and current_rate <= (1 - max_drawdown) * settlement_amount:
        hedged_amount = 0

    if cost_average_factor is not None:
        cost_averaged_amount = hedged_amount * cost_average_factor
        hedged_amount = min(hedged_amount, cost_averaged_amount)

    if dynamic_position_multiplier is not None:
        dynamic_position_size = hedged_amount * dynamic_position_multiplier
        hedged_amount = min(hedged_amount, dynamic_position_size)

    if option_cost is not None:
        hedged_amount -= option_cost

    if stress_testing is not None and stress_testing > 0:
        stress_tested_amount = hedged_amount * stress_testing
        hedged_amount = min(hedged_amount, stress_tested_amount)

    if scenario_analysis is not None and scenario_analysis > 0:
        scenario_analysis_amount = hedged_amount * scenario_analysis
        hedged_amount = min(hedged_amount, scenario_analysis_amount)

    return settlement_amount, hedged_amount


# Example usage
amount_to_hedge = 100000  # Amount to hedge in euros
current_exchange_rate = 1.20  # Current exchange rate (1 euro = 1.20 dollars)
forward_exchange_rate = 1.10  # Forward exchange rate (1 euro = 1.10 dollars)
transaction_cost = 1000  # Transaction cost in dollars
margin_requirement = 0.1  # Margin requirement as a decimal (e.g., 0.1 = 10%)
risk_tolerance = 50000  # Risk tolerance in dollars

settlement_amount, hedged_amount = currency_hedging(
    amount_to_hedge, current_exchange_rate, forward_exchange_rate, transaction_cost, margin_requirement,
    risk_tolerance, stop_loss=1.15, take_profit=1.05, trailing_stop=0.05, partial_hedging_percentage=0.5,
    diversification_factor=0.3, rebalancing_percentage=0.2, correlation_threshold=-0.1,
    volatility_stop_loss=1.1, profit_target=1.3, max_drawdown=0.2, cost_average_factor=0.8,
    dynamic_position_multiplier=1.2, option_cost=500, stress_testing=0.8, scenario_analysis=0.9
)

# Print the results
print("Settlement Amount (USD):", settlement_amount)
print("Hedged Amount (USD):", hedged_amount)

File: currency-hedge.R
currency_hedging <- function(amount, current_rate, forward_rate, transaction_cost, margin_requirement, risk_tolerance,
                            stop_loss=NULL, take_profit=NULL, trailing_stop=NULL, partial_hedging_percentage=NULL,
                            diversification_factor=NULL, rebalancing_percentage=NULL, correlation_threshold=NULL,
                            volatility_stop_loss=NULL, profit_target=NULL, max_drawdown=NULL, cost_average_factor=NULL,
                            dynamic_position_multiplier=NULL, option_cost=NULL, stress_testing=NULL, scenario_analysis=NULL) {
  
  settle <- function(amount, exchange_rate) {
    # Calculates the settlement amount in the target currency based on the amount to be hedged and the exchange rate
    amount * exchange_rate
  }
  
  rate_difference <- forward_rate - current_rate
  
  settlement_amount <- settle(amount, forward_rate)
  
  # Deduct transaction cost from settlement amount
  settlement_amount <- settlement_amount - transaction_cost
  
  # Calculate the margin required based on the hedged amount
  margin_required <- amount * margin_requirement
  
  # Adjust the hedged amount based on the margin requirement
  hedged_amount <- amount + (amount * rate_difference) - margin_required
  
  # Apply risk management strategies
  if (!is.null(stop_loss) && current_rate <= stop_loss) {
    hedged_amount <- 0
  }
  
  if (!is.null(take_profit) && current_rate >= take_profit) {
    hedged_amount <- risk_tolerance
  }
  
  if (!is.null(trailing_stop) && trailing_stop > 0) {
    trailing_stop_loss <- max(current_rate - trailing_stop, stop_loss)
    if (current_rate <= trailing_stop_loss) {
      hedged_amount <- 0
    }
  }
  
  if (!is.null(partial_hedging_percentage)) {
    partial_hedged_amount <- hedged_amount * partial_hedging_percentage
    remaining_hedged_amount <- hedged_amount - partial_hedged_amount
    hedged_amount <- min(hedged_amount, remaining_hedged_amount)
  }
  
  if (!is.null(diversification_factor)) {
    diversification_hedged_amount <- remaining_hedged_amount * diversification_factor
    remaining_hedged_amount <- remaining_hedged_amount - diversification_hedged_amount
    hedged_amount <- min(hedged_amount, remaining_hedged_amount)
  }
  
  if (!is.null(rebalancing_percentage)) {
    rebalancing_amount <- hedged_amount * rebalancing_percentage
    hedged_amount <- hedged_amount - rebalancing_amount
  }
  
  if (!is.null(correlation_threshold)) {
    if (correlation_threshold < 0) {
      # Negative correlation: Reduce position size
      hedged_amount <- hedged_amount * abs(correlation_threshold)
    } else {
      # Positive correlation: Increase position size
      hedged_amount <- hedged_amount * (1 + correlation_threshold)
    }
  }
  
  if (!is.null(volatility_stop_loss) && current_rate - volatility_stop_loss >= stop_loss) {
    hedged_amount <- 0
  }
  
  if (!is.null(profit_target) && current_rate >= profit_target) {
    hedged_amount <- risk_tolerance
  }
  
  if (!is.null(max_drawdown) && max_drawdown > 0 && current_rate <= (1 - max_drawdown) * settlement_amount) {
    hedged_amount <- 0
  }
  
  if (!is.null(cost_average_factor)) {
    cost_averaged_amount <- hedged_amount * cost_average_factor
    hedged_amount <- min(hedged_amount, cost_averaged_amount)
  }
  
  if (!is.null(dynamic_position_multiplier)) {
    dynamic_position_size <- hedged_amount * dynamic_position_multiplier
    hedged_amount <- min(hedged_amount, dynamic_position_size)
  }
  
  if (!is.null(option_cost)) {
    hedged_amount <- hedged_amount - option_cost
  }
  
  if (!is.null(stress_testing) && stress_testing > 0) {
    stress_tested_amount <- hedged_amount * stress_testing
    hedged_amount <- min(hedged_amount, stress_tested_amount)
  }
  
  if (!is.null(scenario_analysis) && scenario_analysis > 0) {
    scenario_analysis_amount <- hedged_amount * scenario_analysis
    hedged_amount <- min(hedged_amount, scenario_analysis_amount)
  }
  
  list(settlement_amount = settlement_amount, hedged_amount = hedged_amount)
}


# Example usage
amount_to_hedge <- 100000  # Amount to hedge in euros
current_exchange_rate <- 1.20  # Current exchange rate (1 euro = 1.20 dollars)
forward_exchange_rate <- 1.10  # Forward exchange rate (1 euro = 1.10 dollars)
transaction_cost <- 1000  # Transaction cost in dollars
margin_requirement <- 0.1  # Margin requirement as a decimal (e.g., 0.1 = 10%)
risk_tolerance <- 50000  # Risk tolerance in dollars

result <- currency_hedging(
  amount_to_hedge, current_exchange_rate, forward_exchange_rate, transaction_cost, margin_requirement,
  risk_tolerance, stop_loss=1.15, take_profit=1.05, trailing_stop=0.05, partial_hedging_percentage=0.5,
  diversification_factor=0.3, rebalancing_percentage=0.2, correlation_threshold=-0.1,
  volatility_stop_loss=1.1, profit_target=1.3, max_drawdown=0.2, cost_average_factor=0.8,
  dynamic_position_multiplier=1.2, option_cost=500, stress_testing=0.8, scenario_analysis=0.9
)

# Print the results
cat("Settlement Amount (USD):", result$settlement_amount, "\n")
cat("Hedged Amount (USD):", result$hedged_amount, "\n")
